src/runtime/inspector_client.cpp

```diff
@@ -311,3 +330,384 @@ void TrInspectorClient::onReset()
   DEBUG(LOG_TAG_ERROR, "Request reset");
   end();
 }
+
+bool TrInspectorClient::tryUpgradeToWebSocket()
+{
+  // Check if this is a WebSocket upgrade request
+  bool hasUpgrade = false;
+  bool hasConnection = false;
+  bool hasWebSocketKey = false;
+  std::string webSocketKey;
+
+  for (const auto &header : headers_)
+  {
+    std::string lowerKey = header.first;
+    std::transform(lowerKey.begin(), lowerKey.end(), lowerKey.begin(), ::tolower);
+
+    if (lowerKey == "upgrade" && header.second == "websocket")
+    {
+      hasUpgrade = true;
+    }
+    else if (lowerKey == "connection")
+    {
+      std::string lowerValue = header.second;
+      std::transform(lowerValue.begin(), lowerValue.end(), lowerValue.begin(), ::tolower);
+      if (lowerValue.find("upgrade") != std::string::npos)
+      {
+        hasConnection = true;
+      }
+    }
+    else if (lowerKey == "sec-websocket-key")
+    {
+      hasWebSocketKey = true;
+      webSocketKey = header.second;
+    }
+  }
+
+  if (hasUpgrade && hasConnection && hasWebSocketKey)
+  {
+    // Check WebSocket connection limit
+    auto inspector = inspector_.lock();
+    if (inspector == nullptr)
+    {
+      return false;
+    }
+
+    if (!inspector->canAcceptWebSocketConnection())
+    {
+      DEBUG(LOG_TAG_INSPECTOR, "WebSocket connection limit reached, rejecting upgrade");
+      // Send 503 Service Unavailable
+      std::stringstream response;
+      response << "HTTP/1.1 503 Service Unavailable\r\n";
```

Comment:
You can add a method for sending 400 or 500 response, which auto calucate the content-length.